internal/driver/deviceclient.go

```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
+	case v2.ValueTypeUint32:
 		var res = binary.BigEndian.Uint32(swap32BitDataBytes(dataBytes, commandInfo.IsByteSwap, commandInfo.IsWordSwap))
-		result, err = sdkModel.NewUint32Value(req.DeviceResourceName, resTime, res)
-	case sdkModel.Uint64:
+		result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeUint32, res)
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
+	case v2.ValueTypeUint64:
 		var res = binary.BigEndian.Uint64(dataBytes)
-		result, err = sdkModel.NewUint64Value(req.DeviceResourceName, resTime, res)
-	case sdkModel.Int16:
+		result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeUint64, res)
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
+	case v2.ValueTypeInt16:
 		var res = int16(binary.BigEndian.Uint16(dataBytes))
-		result, err = sdkModel.NewInt16Value(req.DeviceResourceName, resTime, res)
-	case sdkModel.Int32:
+		result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeInt16, res)
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
+	case v2.ValueTypeInt32:
 		var res = int32(binary.BigEndian.Uint32(swap32BitDataBytes(dataBytes, commandInfo.IsByteSwap, commandInfo.IsWordSwap)))
-		result, err = sdkModel.NewInt32Value(req.DeviceResourceName, resTime, res)
-	case sdkModel.Int64:
+		result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeInt32, res)
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
+	case v2.ValueTypeInt64:
 		var res = int64(binary.BigEndian.Uint64(dataBytes))
-		result, err = sdkModel.NewInt64Value(req.DeviceResourceName, resTime, res)
-	case sdkModel.Float32:
+		result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeInt64, res)
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
+	case v2.ValueTypeFloat32:
 		switch commandInfo.RawType {
-		case sdkModel.Float32:
+		case v2.ValueTypeFloat32:
 			var res = binary.BigEndian.Uint32(swap32BitDataBytes(dataBytes, commandInfo.IsByteSwap, commandInfo.IsWordSwap))
 			var floatResult = math.Float32frombits(res)
-			result, err = sdkModel.NewFloat32Value(req.DeviceResourceName, resTime, floatResult)
-		case sdkModel.Int16:
+			result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeFloat32, floatResult)
+			if err != nil {
+				return nil, err
+			}
+			result.Origin = resTime
+		case v2.ValueTypeInt16:
 			var res = int16(binary.BigEndian.Uint16(dataBytes))
-			result, err = sdkModel.NewFloat32Value(req.DeviceResourceName, resTime, float32(res))
-			driver.Logger.Debug(fmt.Sprintf("According to the rawType %s and the value type %s, convert integer %d to float %v ", INT16, FLOAT32, res, result.ValueToString(contract.ENotation)))
-		case sdkModel.Uint16:
+			result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeFloat32, float32(res))
```

Comment:
Why float32?? If really needed, then use float32 cast on line 194 so this block can be below switch. Same with similar code below.

Reply:
> Why float32??

We provide rawType for converting **integer**  to **float** value , see https://docs.edgexfoundry.org/1.3/examples/Ch-ExamplesModbusdatatypeconversion/#supported-transformations